### DISCUSSION TITLE
Adding xml element with attribute to logback.xml

### DISCUSSION THREAD

**Discussion by MarcoHuigen** [mikefarah/yq#1125] [Q&A] [answered]
Hi all,

Am using yq to adjust YAML and in same project I have to update some xml. yq has xml support... Nice!!!

But, i don't get a simple (??) add to work.... 

logback.xml:
```
<configuration>
  <!-- lotsa appenders and loggers -->
  <root level="INFO">
    <appender-ref ref="SYSTEMLOG"></appender-ref>
    <appender-ref ref="ANOTHERLOG"></appender-ref>
  </root>
</configuration>
```

i want to simply add 
    `<appender-ref ref="YETANOTHERLOG"></appender-ref>`
    
so that I end up with:

```
<configuration>
  <!-- lotsa appenders and loggers -->
  <root level="INFO">
    <appender-ref ref="SYSTEMLOG"></appender-ref>
    <appender-ref ref="ANOTHERLOG"></appender-ref>
    <appender-ref ref="YETANOTHERLOG"></appender-ref>
  </root>
</configuration>   
``` 

Am quite sure the solution must be simple, but after hours of google-search and trying anything and everything (e.g. snippet merge into file ... which also does not function correctly!) my only resolve is to ask the REAL experts :)

Lotsa thx in advance!

P.S. Got remove to work with something like:

`$YQ eval-all -p=xml -o=xml 'del(.configuration.root.appender-ref[] | select ( .+ref == env(YQAPPENDERNAME)) ) ' $fileName > "$tmpFile"`

**Comment by mikefarah** [comment, marked as answer]
Hi - sorry for the late response. 

So when yq parses this xml, 'appender-ref' is an array (as it appears multiple times). All you need to do is add another item to that array.

If it helps (it did for me) first keep the output as yaml (omit the `-o=xml`) to get an understanding of what happening under the hood.

Regardless:

```bash
yq '.configuration.root.appender-ref += {"+ref": "YETANOTHERLOG"}' sample.xml -p=xml  -o=xml
```

**Reply by MarcoHuigen** [reply]
Thx Mike! Had already given up hope to get an answer :).

Will implement this iin my project in stead of "ugly" sed solution I made. Am pretty sure I tried something quite close to your answer, but used "ref" in stead of "+ref". Have / Had difficulty understanding that + sign, but in the end I guess its just a literal...

As said, thx

**Reply by mikefarah** [reply]
The "+" prefix is so it knows it a xml attribute as opposed to a node - see https://mikefarah.gitbook.io/yq/usage/xml for more info

**Comment by austinarbor** [comment]
@mikefarah i'm wondering if maybe there's a regression with this? I am trying to do this with a pom.xml to add a developer to an existing list of developers
```xml
<project>
  <developers>
    <developer>
       <id>existing</id>
    </developer>
  </developers>
</project>
```

when I use your solution
`yq -i '.project.developers += {"+developer" : {"id": "new"}}' pom.xml -p=xml -o=xml`
it includes the literal `+` in the tag
```xml
<project>
  <developers>
    <developer>
      <id>existing</id>
    </developer>
    <+developer>
      <id>new</id>
    </+developer>
</project>
```

If I remove the `+`, it overwrites the existing element
`yq -i '.project.developers += {"developer" : {"id": "new"}}' pom.xml -p=xml -o=xml`
```xml
<project>
  <developers>
    <developer>
      <id>new</id>
    </developer>
  </developers>
</project>
```

any ideas?

**Reply by mikefarah** [reply]
The issue is that `yq` has no way of knowing that developer is an array when it only sees one entry - so it assumes that it's a map (like the rest of the path). What you need to do to append to it, is force the particular entry into an array - which in this case is:

```
yq '.project.developers.developer |= ([.] + {"id": "new"})'
```

```xml
<project>
  <developers>
    <developer>
      <id>existing</id>
    </developer>
    <developer>
      <id>new</id>
    </developer>
  </developers>
</project>
```

**Reply by mikefarah** [reply]
Side note - it can be a little easier to diagnose these issues by outputting to JSON/YAML - as you will explicitly sees what it thinks is a map vs array.

**Reply by austinarbor** [reply]
@mikefarah thanks! your solution is working perfectly

**Comment by code-sergeant** [comment]
I'm running into a similar issue. My use case is that I would like to programmatically add a suppression field to a list of suppressions in an xml file intended for consumption by a Dependency Check tool. 

GIVEN I have an xml file with the following shape:
```
<?xml version="1.0" encoding="UTF-8"?>
<suppressions>
  <suppress>Testing-456</suppress>
  <suppress>Testing-789</suppress>
</suppressions>
```
WHEN I run the following command:
`yq --input-format xml --output-format xml '.suppressions += {"suppress": "CVE-Testing-NEW-VALUE"}'  suppressions.xml`

THEN I get the following incorrect result, where the array is seen as an object field and replaces the key:
```
<?xml version="1.0" encoding="UTF-8"?>
<suppressions>
  <suppress>CVE-Testing-NEW-VALUE</suppress>
</suppressions>
```

This also is broken when you have the correctly formatted json that you want to simply convert to xml:
GIVEN I have some json with the following shape:
```
{
  "+p_xml": "version=\"1.0\" encoding=\"UTF-8\"",
  "suppressions": [
    {
      "suppress": "CVE-Testing-456"
    },
    {
      "suppress": "CVE-Testing-789"
    }
  ]
}
```

WHEN I run the following command:
`yq --input-format json '.suppressions  += {"suppress": "CVE-Testing-NEW-ENTRY"}' -o xml  suppressions.json`

THEN I get the following incorrect result:
```
<?xml version="1.0" encoding="UTF-8"?>
<suppressions>
  <suppress>Testing-456</suppress>
</suppressions>
<suppressions>
  <suppress>Testing-789</suppress>
</suppressions>
<suppressions>
  <suppress>CVE-Testing-NEW-ENTRY</suppress>
</suppressions>
```
WHEN I run this same command, but select a different output format (yml for example), I get the correct array output:
```
+p_xml: version="1.0" encoding="UTF-8"
suppressions:
  - suppress: Testing-456
  - suppress: Testing-789
  - suppress: CVE-Testing-NEW-ENTRY
```

**Reply by code-sergeant** [reply]
I just found a work around to this difference in behaviors. When working in xml, you need to append with the string entry:
`yq -i -p xml -o xml '.suppressions.suppress += "CVE-Testing-NEW-VALUE"' suppressions.xml`

**Reply by mikefarah** [reply]
Yep that's right, 'suppress' is the array as that is the 'root' node that is repeated in the XML
